lib/puppet/pops/serialization/to_data_converter.rb

```diff
@@ -0,0 +1,284 @@
+    #
+    # If `:local_references` is disabled, then this method performs a check for endless recursion before
+    # it yields to the given block. The result of yielding is returned.
+    #
+    # @param value [Object] the value
+    # @yield The block that will produce the data for the value
+    # @return [Data] the result of yielding to the given block, or a hash denoting a reference
+    #
+    # @api private
+    def process(value, &block)
+      if @local_reference
+        id = value.object_id
+        ref = @values[id]
+        if ref.nil?
+          @values[id] = @path.dup
+          yield
+        elsif ref.instance_of?(Hash)
+          ref
+        else
+          json_ref = JsonPath.to_json_path(ref)
+          if json_ref.nil?
+            # Complex key and hence no way to reference the prior value. The value must therefore be
+            # duplicated which in turn introduces a risk for endless recursion in case of self
+            # referencing structures
+            with_recursive_guard(value, &block)
+          else
+            @values[id] = { PCORE_TYPE_KEY => PCORE_LOCAL_REF_SYMBOL, PCORE_VALUE_KEY => json_ref }
+          end
+        end
+      else
+        with_recursive_guard(value, &block)
+      end
+    end
+
+    # Pushes `key` to the end of the path and yields to the given block. The
+    # `key` is popped when the yield returns.
+    # @param key [Object] the key to push on the current path
+    # @yield The block that will produce the returned value
+    # @return [Object] the result of yielding to the given block
+    #
+    # @api private
+    def with(key)
+      @path.push(key)
+      value = yield
+      @path.pop
+      value
+    end
+
+    # @param value [Object] the value to use when checking endless recursion
+    # @yield The block that will produce the data
+    # @return [Data] the result of yielding to the given block
+    def with_recursive_guard(value)
+      id = value.object_id
+      if @recursive_lock
+        if @recursive_lock.include?(id)
+          serialization_issue(Issues::SERIALIZATION_ENDLESS_RECURSION, :type_name => value.class.name)
+        end
+        @recursive_lock[id] = true
+      else
+        @recursive_lock = { id => true }
+      end
+      v = yield
+      @recursive_lock.delete(id)
+      v
+    end
+
+    def unknown_to_data(value)
+      @rich_data ? value_to_data_hash(value) : unknown_to_string_with_warning(value)
+    end
+
+    def unknown_to_string_with_warning(value)
+      str = value.to_s
+      serialization_issue(Issues::SERIALIZATION_UNKNOWN_CONVERTED_TO_STRING, :path => path_to_s, :klass => value.class.name, :value => str)
+      str
+    end
+
+    def non_string_keyed_hash_to_data(hash)
+      if @rich_data
+        to_key_extended_hash(hash)
+      else
+        result = {}
+        hash.each_pair do |key, value|
+          if key.is_a?(Symbol) && @symbol_as_string
+            key = key.to_s
+          elsif !key.is_a(String)
```

Comment:
I created [PR-5951](/puppetlabs/puppet/pull/5951) with a fix and tests since this PR is merged and closed